### PR TITLE
* fix implicit callings explicit constructors

### PR DESCRIPTION
**Pull request by rinse**
Hello, finally I tried to run your emulator :)
Running make gave me some compile errors and I fixed it.

### CODE REVIEW

**Comment by itmo153277** [comment]
Can you create an issue for this one?
What compiler do you use? What is its version?

The solution you propose is not optimal, it would trigger one unnecessary copy constructor. For example, try this:

```c++
#include <iostream>
#include <utility>
#include <map>

struct A {
	explicit A(int x) {
		std::cout << "A(" << x << ')' << std::endl;
	}
	A(const A&) {
		std::cout << "A() copy" << std::endl;
	}
	~A() = default;
};

int main() {
	std::map<int, A> x;
	std::cout << "First try:" << std::endl;
	x.emplace(std::piecewise_construct, std::forward_as_tuple(0), std::forward_as_tuple(0));
	std::cout << "Second try:" << std::endl;
	x.emplace(1, A(1));
	return 0;
}
```

It will output:
```
First try:
A(0)
Second try:
A(1)
A() copy
```

File: include/vpnes/core/bus.hpp [hunk 1]
```diff
@@ -362,7 +362,7 @@ class CBus {
 	template <class T>
 	void addPreReadHook(std::uint16_t addr, T *device,
 	    typename CAddrHookMapped<T>::addrHook_t hook) {
-		m_ReadHooksPre.emplace(addr, new CAddrHookMapped<T>(device, hook));
+		m_ReadHooksPre.emplace(addr, std::unique_ptr<CAddrHook>(new CAddrHookMapped<T>(device, hook)));
```
Comment (itmo153277) [inline include/vpnes/core/bus.hpp:365]: `std::is_constructible<std::unique_ptr<CAddrHook>, CAddrHookMapped<T> *>::value` should be `true`. Is it a compiler issue?

File: include/vpnes/core/device.hpp [hunk 2]
```diff
@@ -520,7 +520,7 @@ class CEventDevice : public CClockedDevice {
 	 */
 	void registerDeviceEvent(CEvent *event) {
 		assert(m_EventData.find(event) == m_EventData.end());
-		auto newData = m_EventData.emplace(event, event);
+		auto newData = m_EventData.emplace(event, SEventData(event));
```
Comment (itmo153277) [inline include/vpnes/core/device.hpp:523]: `std::is_constructible<SEventData, CEvent *>::value` should definitely be `true`.

File: include/vpnes/core/device.hpp [hunk 3]
```diff
@@ -643,7 +643,7 @@ class CEventManager {
 		assert(eventMap.find(name) == eventMap.end());
 		typename T::CEvent *event = new typename T::template CLocalEvent<T>(
 		    name, time, enabled, device, std::forward<TArgs>(args)...);
-		events.emplace(device, event);
+		events.emplace(device, std::unique_ptr<typename T::CEvent>(event));
```
Comment (itmo153277) [inline include/vpnes/core/device.hpp:646]: Same as above, `T::CEvent *` can be converted to `CDeviceEvent *`, so it should be constructible

**Comment by itmo153277** [comment]
After some research I discovered that this was not a part of c++11 and was introduced only in c++17. So the solution would be upgrading to c++17 or to disable [this rule](https://google.github.io/styleguide/cppguide.html#Implicit_Conversions) in the project and remove `explicit` modifier.

**Comment by itmo153277** [comment]
After some thinking on this issue, I decided:
- the code should be upgraded to c++17,
- parts with emplacing `std::unique_ptr` should be rewritten, it is a bad style and can lead to memory leaks.

I will try to do this tomorrow along with other changes. If you don't have any objections, I will close this PR.

**Comment by rinse** [comment]
I have no objection against your suggestion. I'm sorry for lack of information but yes, the constructor of `std::pair`, `pair(U1&&, U2&&)` requires that each `U1` and `U2` is implicitly convertible to `T1` and `T2`, the types of `first` and `second` until C++17.

I don't think it not necessary to make an issue now, do I ?

**Pull request closed without merging**